ARCHITECTURE.md

```diff
@@ -0,0 +1,122 @@
+
+This synthesized module is never persisted to disk.[^debug] Instead, it is presented to the TS compiler, either via a programmatically constructed compiler instance for command line runs or via a language server instance for code editor operations. The results from the TS check are then mapped back to the original template provide diagnostics in the correct location, and sometimes further massaged or rewritten to make them make sense in terms of templates rather than the TS representation of templates.
+
+[^debug]: It is possible to write to disk with a debug flag, but the statement holds for all normal operations.
+
+## Code structure
+
+Glint is composed of the following major components:
+
+
+### Config
+
+**Role:** a library which handles interpreting Glint configuration: which modules should be part of the transform, and how should they be interpreted?
+
+**Invariants:** should know nothing about the rest of the pipeline; it only needs to understand how to parse a Glint configuration and hand it off to the rest of the pipeline.
+
+**Home:** `packages/config`
+
+**Package name:** `@glint/config`
+
+
+### Template DSL
+
+**Role:** a set of TypeScript type definitions which Glint uses as the TypeScript representation of the constituent elements of a template: content, elements, components, expressions, `...attributes`, modifiers, and templates themselves (with or without backing class contexts).
+
+The DSL defines three broad things:
+
+- What it means to *emit* some template content, i.e. what the semantics of emitting `<div {{modifier}}>` are as a kind of TS function application.
+
+- How to *resolve* some template element: when we see `{{this.foo}}`, is that a value to "emit" directly, or is it actually a no-argument function invocation?
+
+- A host of type definitions. Some of these are internal and used to make *emit* and *resolve* work, while others are public types, usable by consumers for authoring their apps. These definitions includes the definitions for types authors can use for writing things like partially applied components.
+
+Note: The fact that the synthesized module is *not* persisted to disk means that some things you might expect to be useful navigation tools within the repo, like the TypeScript *Find All References* command, will not find any references to these.
+
+**Invariants:** ==TODO==
+
+**Home:** `packages/template`
+
+**Package name:** `@glint/template`
+
+
+### Environments
+
+**Role:** layers on top of the **template DSL** to define items available in global scope as well as per-environment definitions for what the API is for different kinds of items—basically, a type-level implementation of Glimmer/Ember’s runtime idea of “managers,” which allow a different surface API for the same underlying primitives of components, helpers, etc. A **config** specifies the **environment** to use.
+
+**Invariants:** ==TODO==
```

Comment:
```suggestion
**Invariants:**

- Environments have one entrypoint for their configuration (specified by the `glint-environment` key in `package.json`), and one or more types-only entrypoints where they expose their specialized **template DSL** implementation(s).[^environments-dsl]
- An environment will typically depend on the bedrock **template DSL** as a basis for its own DSL implementation, and it might reference types from the **config** package in defining its configuration, but an environment should have no reason to ever execute code from other Glint packages.

[^environments-dsl]: I don't think any environments have multiple entrypoints today, but a hypothetical unified Ember environment in the near future would have separate DSLs for loose-mode and strict-mode templates.
```